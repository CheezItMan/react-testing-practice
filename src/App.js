import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import { Button } from 'react-bootstrap';
import Gift from './components/gift';
import { max_number } from './helpers';


class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      gifts: [],
    }
  }

  addGift = () => {
    const { gifts } = this.state;
    const ids = this.state.gifts.map(gift => gift.id);

    gifts.push({
      id: max_number(ids) + 1,
    });

    this.setState(gifts: gifts);
  }

  removeGift = id => {
    const gifts = this.state.gifts.filter(gift => gift.id != id );

    this.setState({gifts})
  }

  giftList = () => {
    return (
      <section className="gift-list">
      {
        this.state.gifts.map(gift => {
          return (
            <Gift
              key={gift.id}
              gift={gift}
              removeGift={this.removeGift}
            />
          )
        })
      }
      </section>
    );
  }


  render() {
    return (
      <div className="App">
        <header><h2>Gift Giver</h2></header>
        <main>
          <Button className="btn-add" onClick={this.addGift} >Add Gift</Button>
          {this.giftList()}
        </main>
      </div>
    );
  }
}

export default App;
