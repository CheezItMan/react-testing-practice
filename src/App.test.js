import React from 'react';
// Allows you to shallowly render the component (only 1-layer deep)
import { shallow } from 'enzyme';
import App from './App';



describe('App', () => {
  let app;

  beforeEach(() => {
    app = shallow(<App />);
  });

  test('it renders correctly', () => {
    expect(app).toMatchSnapshot();
  });

  test('it initializes the `state` with an empty list of gifts', () => {
    expect(app.state().gifts).toEqual([]);
  });

  describe('When clicking the add-gift button', () => {
    const id = 1;
    beforeEach(() => {
      app.find('.btn-add').simulate('click');

    });

    test('it adds new gifts to `state`', () => {
      expect(app.state().gifts.length).toEqual(1);
      expect(app.state().gifts).toEqual([{ id }]);
    });

    test('it adds a new gift to the giftlist', () => {
      expect(app.find('.gift-list').children().length).toEqual(1);
    });

    test('It creates a gift component', () => {
      expect(app.find('Gift').exists()).toBe(true);
    });



    describe('and the user wants to remove the added gift', () => {
      beforeEach(() => {
        app.instance().removeGift(id);
      });

      test('it removes the gift from `state`', () => {
        expect(app.state().gifts).toEqual([]);
      });
    });
  });

});
