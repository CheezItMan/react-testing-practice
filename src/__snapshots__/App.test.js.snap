// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App it renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header>
          <h2>
                    Gift Giver
          </h2>
</header>,
        <main>
          <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="default"
                    className="btn-add"
                    disabled={false}
                    onClick={[Function]}
          >
                    Add Gift
          </Button>
          <section
                    className="gift-list"
          >
                    
          </section>
</main>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h2>
            Gift Giver
</h2>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Gift Giver",
          },
          "ref": null,
          "rendered": "Gift Giver",
          "type": "h2",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              className="btn-add"
              disabled={false}
              onClick={[Function]}
>
              Add Gift
</Button>,
            <section
              className="gift-list"
>
              
</section>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": false,
              "block": false,
              "bsClass": "btn",
              "bsStyle": "default",
              "children": "Add Gift",
              "className": "btn-add",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add Gift",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "gift-list",
            },
            "ref": null,
            "rendered": Array [],
            "type": "section",
          },
        ],
        "type": "main",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header>
            <h2>
                        Gift Giver
            </h2>
</header>,
          <main>
            <Button
                        active={false}
                        block={false}
                        bsClass="btn"
                        bsStyle="default"
                        className="btn-add"
                        disabled={false}
                        onClick={[Function]}
            >
                        Add Gift
            </Button>
            <section
                        className="gift-list"
            >
                        
            </section>
</main>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h2>
              Gift Giver
</h2>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Gift Giver",
            },
            "ref": null,
            "rendered": "Gift Giver",
            "type": "h2",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="default"
                className="btn-add"
                disabled={false}
                onClick={[Function]}
>
                Add Gift
</Button>,
              <section
                className="gift-list"
>
                
</section>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "active": false,
                "block": false,
                "bsClass": "btn",
                "bsStyle": "default",
                "children": "Add Gift",
                "className": "btn-add",
                "disabled": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add Gift",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
                "className": "gift-list",
              },
              "ref": null,
              "rendered": Array [],
              "type": "section",
            },
          ],
          "type": "main",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
